[server] Restrict snapshot access based on repository access

jankeromnes opened the pull request:
## Description
<!-- Describe your changes in detail -->

Restrict snapshot access based on repository access.

Also refactors:
- `GuardedSnapshot.workspace` (simplified)
- `WorkspaceLogAccessGuard` → `RepositoryResourceGuard`
- `RepositoryService.canAccessHeadlessLogs` → `RepositoryProvider.hasReadAccess`

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/8257

## How to test
<!-- Provide steps to test this PR -->

1. Create a snapshot for a private repo
2. Verify that you can open the snapshot
3. Verify that other members of the private repo can open the snapshot
4. Verify that non-members of the private repo can **not** open the snapshot

Also:

1. Create a snapshot for a public repo
2. Verify that anyone can open the snapshot (even if they haven't connected the repo's Git provider)

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
Restrict snapshot access based on repository access
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

jankeromnes commented on components/server/src/gitlab/gitlab-repository-provider.ts:
```diff
@@ -99,4 +99,18 @@ export class GitlabRepositoryProvider implements RepositoryProvider {
         return [];
     }
+
+    async hasReadAccess(user: User, owner: string, repo: string): Promise<boolean> {
+        try {
+            // If we can "see" a project we are allowed to read it
+            const api = await this.gitlab.create(user);
+            const response = await api.Projects.show(`${owner}/${repo}`);
+            if (GitLab.ApiError.is(response)) {
+                return false;
+            }
+            return true;
+        } catch (err) {
+            return false;
+        }
+    }
```
FIXME: This also fails for public GitLab repositories when you have no connected GitLab identity:

```
Oh, no! Something went wrong!
operation not permitted: missing get permission on snapshot
Unknown Error: { "code": 403 }
```

jankeromnes commented:
Many thanks!

Holding because I'd like to see if I can detect the missing integration, and at least show a "Connect" button on the error screen in that case.

/hold

geropl commented on components/server/src/github/github-repository-provider.ts:
```diff
@@ -128,4 +128,21 @@ export class GithubRepositoryProvider implements RepositoryProvider {
         return (result.data.viewer?.repositoriesContributedTo?.edges || []).map((edge: any) => edge.node.url)
     }
+
+    async hasReadAccess(user: User, owner: string, repo: string): Promise<boolean> {
+        try {
+            // If you have no "viewerPermission" on a repository you may not read it
+            // Ref: https://docs.github.com/en/graphql/reference/enums#repositorypermission
+            const result: any = await this.githubQueryApi.runQuery(user, `
+                query {
+                    repository(name: "${repo}", owner: "${owner}") {
+                        viewerPermission
+                    }
+                }
+            `);
+            return result.data.repository !== null;
+        } catch (err) {
+            return false;
+        }
+    }
```
You mean if you don't have a GitHub account registered with Gitpod? Yes, that's expected. Would be nice to fix that, though. :+1:

geropl replied:
But maybe that's a separate problem. I like @jldec suggestions: https://github.com/gitpod-io/gitpod/pull/8306#issuecomment-1044526336

AlexTugarev replied:
@jankeromnes, that's ok here if we can assume that it's already checked. see https://github.com/gitpod-io/gitpod/pull/8306/files#diff-b4ab99413da84ca8b9ecc7d07a462a1eb0a7bc086cbb90b90d5b59c2b170eb44R490 there is a detailed error thrown at the frontend which can be consumed properly.

geropl reviewed (approved):
Code LGTM as well

I like the WorkspaceLogAccessGuard → RepositoryResourceGuard refactoring :+1:

jankeromnes commented:
Left to do:
- [ ] Understand and maybe resolve the snapshot bugs mentioned above
- [ ] Surface the message `Snapshot URLs require read access to the underlying repository. Please request access from the repository owner.` in the UI when snapshot access fails

jankeromnes commented:
Hmm, I've added some debug logs, but could not reproduce the errors above. I guess this works as expected.

I've also added a more specific error message when you don't have repo access:

**Integration connected, repo access**

Opening the snapshot works ✅ 

**Integration connected, no repo access**

<img width="378" alt="Screenshot 2022-02-19 at 14 40 13" src="https://user-images.githubusercontent.com/599268/154803379-caa1ee73-0a35-4605-b3b6-d984d779bd22.png">

**Integration not connected (can't verify access)**

<img width="390" alt="Screenshot 2022-02-19 at 14 37 11" src="https://user-images.githubusercontent.com/599268/154803397-525ae595-4e8f-4cf8-9abd-59440393e5aa.png">

I feel like we could still make these error pages look nicer somehow.

AlexTugarev commented on components/server/src/bitbucket-server/bitbucket-server-repository-provider.ts:
```diff
@@ -100,4 +96,13 @@ export class BitbucketServerRepositoryProvider implements RepositoryProvider {
+    }
+
+    hasReadAccess(user: User, owner: string, repo: string): Promise<boolean> {
+        throw new Error("Method not implemented.");
```
in that case `return true` would be a better default value, I guess. even if not correct in all cases, it should be stopper in general.

jankeromnes replied:
Good point, thanks @AlexTugarev. But `return true` would mean "allow anyone to open snapshots of Bitbucket server repositories that are potentially private", which goes somewhat against the intent of this change (i.e. restrict snapshot access to only users who are allowed to read the repository).

Then again, if you have an account for a particular Bitbucket server, maybe it means you should have access to all repos/snapshots? (E.g. maybe these Bitbucket servers are for companies, and only employees can have an account there?)

AlexTugarev replied:
Do I read the code right, so that I see no way to open a workspace any more for BB Server if the `Method not implemented` error remains in place? If so, please change to the default `true`, and add a `TODO` in there. We're not expecting to make it feature complete very soon.

BB Server support is an alpha feature for Gitpod Self-Hosted only.

Also, the nature of the calling site is to check for a relevant git connection first, so this is not big issue. 

> goes somewhat against the intent of this Pull Request

I agree and understand. Let's call it an exception and maybe put a short `TODO` comment in there. That BB Server support is alpha, so ... 🤷🏻‍♂️

jankeromnes replied:
> Do I read the code right, so that I see no way to open a workspace any more for BB Server if the `Method not implemented` error remains in place?

Only when you're trying to open a Snapshot made on a Bitbucket server repo.

Creating regular workspaces or prebuilds for Bitbucket server repos continues to work as before.

However, now that you mention it, `hasReadAccess` is also called when trying to watch headless logs. 🤔 I wonder what happens in that case. Maybe returning `false` is safer here.

jankeromnes replied:
Or, actually, returning `true` and calling that an exception as you suggested.

It's a small security concern, but going from:
- snapshots can be opened by anyone

to:
- snapshots are restricted to repo readers, except for Bitbucket servers where they're still unrestricted

is already a good first step I think.

AlexTugarev commented on components/server/src/repohost/repository-provider.ts:
```diff
@@ -14,4 +14,5 @@ export interface RepositoryProvider {
     getBranches(user: User, owner: string, repo: string): Promise<Branch[]>;
     getCommitInfo(user: User, owner: string, repo: string, ref: string): Promise<CommitInfo | undefined>;
     getUserRepos(user: User): Promise<string[]>;
+    hasReadAccess(user: User, owner: string, repo: string): Promise<boolean>;
```
👍🏻

AlexTugarev reviewed (approved):
LGTM

jldec commented:
This looks good now @jankeromnes - thanks for improving the error message.

jldec reviewed (approved):
merci bien

jankeromnes commented:
Many thanks for the reviews! 🙏

Releasing the [hold-down clamps](https://www.youtube.com/watch?v=K7akDwTjRpI), go for launch 🚀

/unhold
